rust/src/meas/geometric/mod.tex

```diff
@@ -0,0 +1,283 @@
+        \item Variable \texttt{bounds} must be of type \texttt{Option(D::Atom, D::Atom)}
+        \item Type \texttt{D} must have trait \texttt{GeometricDomain}
+        \item Type \texttt{D::Atom} must have traits \texttt{TotalOrd} and \texttt{InfCast<QO>}
+        \item Type \texttt{QO} must have traits \texttt{float} and \texttt{DistanceConstant<D::Atom>} 
+        \item Type \texttt{f64} must have trait \texttt{InfCast<QO>}
+        \item Type \texttt{T} must have traits \texttt{SampleTwoSidedGeometric} and \texttt{CheckNull}
+    \end{itemize}
+\end{itemize}
+
+\vicki{not sure about the domain/T stuff; it's mentioned in the implementation of the domain so I figured should also have it here} 
+
+\subsubsection*{Postconditions}
+
+\begin{itemize}
+    \item A \texttt{Measurement} is returned (i.e., if a \texttt{Measurement} cannot be returned successfully, then an error should be returned).
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_base_geometric(scale, bounds={"lower": T_MIN, "upper": T_MAX}):
+    input_domain = AllDomain(T)
+    output_domain = AllDomain(T)
+    input_metric = L1Distance(T)
+    similarity_metric = MaxDivergence(QO)
+    
+    # check scale sign
+    if (scale < 0): |\label{line:scale}|
+        raise Exception("scale must not be negative")
+        
+    # check bounds - can discuss if python dict is best representation of this
+    if (bounds["lower"] > bounds["upper"]): |\label{line:bounds}|
+        raise Exception("lower may not be greater than upper")
+
+    def privacy_map(d_in: u_32) -> float: |\label{line:map}|
+        if (d_in < 0): |\label{line:positivedin}|
+            raise Exception("sensitivity must be non-negative")
+        if (scale == 0):
+            return QO_MAX
+        return d_in * inf_div(scale);
+
+    # these are two separate functions in original, but can discuss if we want to combine the two
+    def noise_function(arg : T, scale : QO, bounds : Optional[{T, T}]) -> Function<f64, Optional[{T, T}]>: |\label{line:noise_fn}|
+        return function(inf_cast(scale), bounds)
+        
+    def function(arg : T, scale : f64, bounds: Optional[{T, T}]) -> <D, QO>: |\label{line:fn}|
+        sample_two_sided_geometric(arg, scale, bounds)
+    
+    return Measurement(input_domain, output_domain, function, input_metric, similarity_metric, privacy_map)
+
+\end{lstlisting}
+\hanwen{Changed the input metric to L1Distance for vector
+    (was AbsoluteDistance, but I guess AbsoluteDistance could 
+    be viewed as a special case of L1Distance with dimension 1)}
+\hanwen{re. noise function, I guess it's defined separately to account for different input domain (AllDomain and VectorDomain), but everything else looks pretty much the same to me. I would vote for using just one function and call
+sample\_two\_sided\_geometric once inside, plus some brief explanation of the domain?}
+
+\vicki{may be floating point issues with \texttt{is\_sign\_negative}}
```

Comment:
I don't anticipate floating-point issues here. Checking if the sign is negative essentially amounts to checking if the first bit is set, which is always exact. I'm also ok with rejecting -0. when -0. could technically still be ok.